micro-rdk/src/common/ota.rs

```diff
@@ -0,0 +1,359 @@
+    project_name: [u8; 32],
+    /// compile time
+    time: [u8; 16],
+    /// compile date
+    date: [u8; 16],
+    idf_ver: [u8; 32],
+    app_elf_sha256: [u8; 32],
+    /// minimal eFuse block revision supported by image, in format: major * 100 + minor
+    min_efuse_blk_rev_full: u16,
+    /// maximal eFuse block revision supported by image, in format: major * 100 + minor
+    max_efuse_blk_rev_full: u16,
+    reserv2: [u32; 19],
+}
+
+// TODO(RSDK-9214)
+#[derive(Error, Debug)]
+pub enum OtaError {
+    #[error("{0}")]
+    ConfigError(String),
+    #[cfg(feature = "esp32")]
+    #[error(transparent)]
+    EspError(#[from] EspError),
+    #[error("failed to initialize ota process")]
+    InitError,
+    #[error("new image size is invalid: {0} bytes")]
+    InvalidImageSize(usize),
+    #[error("{0}")]
+    Other(String),
+}
+
+#[cfg(feature = "esp32")]
+type OtaConnector = crate::esp32::tcp::Esp32H2Connector;
+#[cfg(not(feature = "esp32"))]
+type OtaConnector = crate::native::tcp::NativeH2Connector;
+
+pub(crate) struct OtaService {
+    exec: Executor,
+    connector: OtaConnector,
+    url: String,
+}
+
+impl OtaService {
+    pub(crate) fn from_config(
+        ota_config: &ServiceConfig,
+        cert: TlsCertificate,
+        exec: Executor,
+    ) -> Result<Self, OtaError> {
+        // TODO(RSDK-9205)
+        let kind: Kind = ota_config
+            .attributes
+            .as_ref()
+            .ok_or_else(|| OtaError::ConfigError("config missing `attributes`".to_string()))?
+            .fields
+            .get("url")
+            .ok_or(OtaError::ConfigError(
+                "config missing `url` field".to_string(),
+            ))?
+            .kind
+            .as_ref()
+            .ok_or(OtaError::ConfigError(
+                "failed to get inner `Value`".to_string(),
+            ))?
+            .try_into()
+            .map_err(|e: AttributeError| OtaError::ConfigError(e.to_string()))?;
+
+        let url = match kind {
+            Kind::StringValue(s) => s,
+            _ => {
+                return Err(OtaError::ConfigError(format!(
+                    "invalid url value: {:?}",
+                    kind
+                )))
+            }
+        };
+
+        let mut connector = OtaConnector::default();
+
+        connector.set_server_certificates(cert.certificate.clone(), cert.private_key.clone());
+
+        Ok(Self {
+            url,
+            connector,
+            exec,
+        })
+    }
+
+    pub(crate) async fn update(&mut self) -> Result<(), OtaError> {
+        let mut uri = self
+            .url
+            .parse::<hyper::Uri>()
+            .map_err(|_| OtaError::ConfigError(format!("invalid url: {}", self.url)))?;
+        if uri.port().is_none() {
+            if uri.scheme_str() != Some("https") {
+                log::error!("no port found and not https");
+            }
+
+            let mut auth = uri.authority().expect("str").to_string();
+            auth.push_str(":443");
```

Comment:
Does this represent a port? What determines this value?

Reply:
this represents a port. This is specifically in the case that the scheme is `https` and no port is specified. `443` is the known port for https (connection attempt will fail without it). hyper doesn't make any assumptions for you so you need to handle the case that a port isn't included.

If you include a port in the url (ex `https://my_url.com:9000`) it will use that port instead.

Reply:
ok, sounds good

Reply:
neither hyper nor http crates define the https default port, I'm guessing because these crates try to minimize assumptions. we could look at another dependency, but I'd rather leave this as-is and add a comment.

note: `reqwest` handles it in-place, [link](https://github.com/search?q=repo%3Aseanmonstar%2Freqwest%20443&type=code)